src/Transfer/Transfer.php

```diff
@@ -0,0 +1,189 @@
+        $this->destination->registerTransferCache($this->resourceCache);
+        $this->destination->setSource($source);
+
+        return $this;
+    }
+
+    /**
+     * @var Source
+     */
+    protected Source $source;
+
+    /**
+     * @var Destination
+     */
+    protected Destination $destination;
+
+    /**
+     * @var string
+     */
+    protected string $currentResource;
+
+    /**
+     * A local cache of resources that were transferred.
+     *
+     * @var ResourceCache
+     */
+    protected ResourceCache $resourceCache;
+
+    /**
+     * @var array
+     */
+    protected array $options = [];
+
+    /**
+     * @var array
+     */
+    protected array $callbacks = [];
+
+    /**
+     * @var array
+     */
+    protected array $events = [];
+
+    public function getStatusCounters()
+    {
+        $status = [];
+
+        foreach ($this->resourceCache->getAll() as $resources) {
+            foreach ($resources as $resource) {
+                /** @var Resource $resource */
+                if (!array_key_exists($resource->getName(), $status)) {
+                    $status[$resource->getName()] = [
+                        Resource::STATUS_PENDING => 0,
+                        Resource::STATUS_SUCCESS => 0,
+                        Resource::STATUS_ERROR => 0,
+                        Resource::STATUS_SKIPPED => 0,
+                        Resource::STATUS_PROCESSING => 0,
+                        Resource::STATUS_WARNING => 0,
+                    ];
+                }
+
+                $status[$resource->getName()][$resource->getStatus()]++;
+            }
+        }
+
+        return $status;
+    }
+
+    /**
+     * Transfer Resources between adapters
+     *
+     * @param array $resources
+     * @param callable $callback (array $resources)
+     */
+    public function run(array $resources, callable $callback): void
+    {
+        // Allows you to push entire groups if you want.
+        $computedResources = [];
+        foreach ($resources as $resource) {
+            if (is_array($resource)) {
+                $computedResources = array_merge($computedResources, $resource);
+            } else {
+                $computedResources[] = $resource;
+            }
+        }
+
+        $this->destination->run($computedResources, $callback, $this->source);
+    }
+
+    /**
+     * Get Resource Cache
+     *
+     * @return ResourceCache
+     */
+    public function getResourceCache(): ResourceCache
+    {
+        return $this->resourceCache;
+    }
+
+    /**
+     *  Get Current Resource
+     *
+     * @return string
+     **/
+
+    public function getCurrentResource(): string
+    {
+        return $this->currentResource;
+    }
+
+    /**
+     * Get Transfer Report
+     *
+     * @param string $statusLevel
+     * If no status level is provided, all status types will be returned.
+     *
+     * @return array
+     */
+    public function getReport(string $statusLevel = ''): array
```

Comment:
It's always better to stick to primitives, but if not possible, we might want to consider a Report class that just have simpler setters and getters instead of an unstructured assoc array.